torchmetrics/functional/audio/si_sdr.py

```diff
@@ -0,0 +1,64 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import torch
+from torch import Tensor
+
+
+def si_sdr(preds: Tensor, target: Tensor, zero_mean: bool = False) -> Tensor:
+    """ scale-invariant signal-to-distortion ratio (SI-SDR)
+
+    Args:
+        preds:
+            shape [..., time]
+        target:
+            shape [..., time]
+        zero_mean:
+            if to zero mean target and preds or not
+
+    Raises:
+        TypeError:
+            if target and preds have a different shape
+
+    Returns:
+        si-sdr value of shape [...]
+
+    Example:
+        >>> from torchmetrics.functional.audio import si_sdr
+        >>> target = torch.tensor([3.0, -0.5, 2.0, 7.0])
+        >>> preds = torch.tensor([2.5, 0.0, 2.0, 8.0])
+        >>> si_sdr_val = si_sdr(preds, target)
+        >>> si_sdr_val
+        tensor(18.4030)
+
+    References:
+        [1] Le Roux, Jonathan, et al. "SDR half-baked or well done." IEEE International Conference on Acoustics, Speech
+         and Signal Processing (ICASSP) 2019.
+    """
+
+    if target.shape != preds.shape:
+        raise TypeError(f"Inputs must be of shape [..., time], got {target.shape} and {preds.shape} instead")
```

Comment:
The error message is not consistent with the check.
Please just use this function: https://github.com/PyTorchLightning/metrics/blob/90743f0240906e61a715832b8c85a2d71ea57b19/torchmetrics/utilities/checks.py#L23-L26
(maybe you can improve it by including the printing of the shapes as your version does)

Reply:
It seems that after changing this function, we need to refactor many test files, for example:

```
def test_error_on_different_shape(metric_class=ExplainedVariance):
    metric = metric_class()
    with pytest.raises(RuntimeError, match='Predictions and targets are expected to have the same shape'):
        metric(torch.randn(100, ), torch.randn(50, ))
```

So, I just leave this function unchanged.
And, I replaced my check code with `_check_same_shape` function.